etc/keepalived/kp_local_loop.py

```diff
@@ -0,0 +1,665 @@
+        if command == SYSTEMCTL_STATUS:
+            return 'up'
+    except subprocess.CalledProcessError as e:
+        print("Failed to %(cmd)s %(srvc)s service: "
+              "%(err)s %(out)s", {'cmd': command, 'srvc': service,
+                                  'err': e, 'out': e.output})
+        if command == SYSTEMCTL_STATUS:
+            return 'down'
+
+
+def parse_isotime(timestr):
+    try:
+        return iso8601.parse_date(timestr)
+    except iso8601.ParseError as e:
+        raise ValueError(six.text_type(e))
+    except TypeError as e:
+        raise ValueError(six.text_type(e))
+
+
+def ping(ipaddr):
+    cli = ['ping', '-c1', '-w1']
+    cli.append(ipaddr)
+    proc = subprocess.Popen(cli,
+                            stdout=subprocess.PIPE,
+                            stderr=subprocess.PIPE)
+    proc.communicate()
+
+    return proc.returncode == 0
+
+
+class Service(object):
+    def __init__(self, name, alarmed, alarmed_at, restarted, restarted_at,
+                 is_necessary, node_role, node_type, status, **kwargs):
+        self.name = name
+        self.alarmed = alarmed
+        self.alarmed_at = alarmed_at
+        self.restarted = restarted
+        self.restarted_at = restarted_at
+        self.is_necessary = is_necessary
+        self.node_role = node_role
+        self.node_type = node_type
+        self.status = status
+
+    def is_abnormal(self):
+        cur_status = service_check(SYSTEMCTL_STATUS, self.name,
+                                   self.node_role, self.node_type)
+        if self.status != cur_status:
+            self.status = cur_status
+            zk_cli = get_zk_cli()
+            kwargs = {}
+            if self.restarted and self.status == 'up':
+                kwargs['restarted'] = None
+                kwargs['restarted_at'] = None
+                self.restarted = None
+                self.restarted_at = None
+            zk_cli.update_service(self.name, self.node_role, self.node_type,
+                                  status=self.status, **kwargs)
+        return self.status != 'up'
+
+    def is_alarmed_timeout(self):
+        over_time = parse_isotime(
+            self.alarmed_at) + datetime.timedelta(
+            days=2)
+        current_time = datetime.datetime.utcnow().replace(tzinfo=iso8601.UTC)
+        return current_time > over_time
+
+    def post_alarmed_if_possible(self):
+        if not self.alarmed:
+            zk_cli = get_zk_cli()
+            updated_svc = zk_cli.update_service(
+                self.name, self.node_role, self.node_type, alarmed=True)
+            self.alarmed = True
+            self.alarmed_at = updated_svc['alarmed_at']
+
+    def post_restarted_if_possible(self):
+        if not self.alarmed:
+            zk_cli = get_zk_cli()
+            updated_svc = zk_cli.update_service(
+                self.name, self.node_role, self.node_type,
+                restarted=True, status='restarting')
```

Comment:
I didn't find the code to set 'restarting' to 'up' or 'down'?

Reply:
Please check in L145， each loop， if the current restart is different with restarting, it will store the new status in zk.

Reply:
Emm, right. So you **update** the service in `is_abnormal` function which I assume is a function for **check**. It's a little strange.